flink-state-backends/flink-statebackend-heap-spillable/src/main/java/org/apache/flink/runtime/state/heap/CopyOnWriteSkipListStateMap.java

```diff
@@ -0,0 +1,1527 @@
+					doPhysicalRemove(currentNode, prevNode, nextNode);
+					logicallyRemovedNodes.remove(currentNode);
+					totalSize--;
+					deleteCount++;
+					currentNode = nextNode;
+					continue;
+				}
+
+				c = compareByteBufferAndNode(keyByteBuffer, keyOffset, keyLen, currentNode);
+
+				if (c > 0) {
+					prevNode = currentNode;
+					currentNode = nextNode;
+					continue;
+				}
+
+				if (c == 0) {
+					int version = SkipListUtils.helpGetNodeLatestVersion(currentNode, spaceAllocator);
+					boolean needCopyOnWrite = version < highestRequiredSnapshotVersion;
+					long oldValuePointer;
+
+					if (needCopyOnWrite) {
+						oldValuePointer = updateValueWithCopyOnWrite(currentNode, value);
+					} else {
+						oldValuePointer = updateValueWithReplace(currentNode, value);
+					}
+
+					byte oldStatus = helpSetNodeStatus(currentNode, SkipListUtils.NodeStatus.PUT.getValue());
+					if (oldStatus == SkipListUtils.NodeStatus.REMOVE.getValue()) {
+						logicallyRemovedNodes.remove(currentNode);
+					}
+
+					S oldState = null;
+					if (returnOldState) {
+						oldState = helpGetState(oldValuePointer);
+					}
+
+					// for the replace, old value space need to free
+					if (!needCopyOnWrite) {
+						spaceAllocator.free(oldValuePointer);
+					}
+
+					return oldState;
+				}
+			}
+
+			// if current node is NIL_NODE or larger than the key, a new node will be inserted
+			break;
+		}
+
+		int level = getRandomIndexLevel();
+		levelIndexHeader.updateLevel(level);
+
+		int totalMetaKeyLen = SkipListUtils.getKeyMetaLen(level) + keyLen;
+		long node = this.spaceAllocator.allocate(totalMetaKeyLen);
+
+		int totalValueLen = SkipListUtils.getValueMetaLen() + value.length;
+		long valuePointer = spaceAllocator.allocate(totalValueLen);
+
+		doWriteKey(node, level, keyByteBuffer, keyOffset, keyLen, valuePointer, currentNode);
+		doWriteValue(valuePointer, value, stateMapVersion, node, NIL_VALUE_POINTER);
+
+		helpSetNextNode(prevNode, node, 0);
+
+		if (level > 0) {
+			SkipListUtils.buildLevelIndex(node, level, keyByteBuffer, keyOffset, levelIndexHeader, spaceAllocator);
+		}
+
+		totalSize++;
+
+		return null;
+	}
+
+	/**
+	 * Remove the key from the skip list. The key can be removed logically or physically.
+	 * Logical remove means put a null value whose size is 0. If the key exists before,
+	 * the old value state will be returned.
+	 *
+	 * @param keyByteBuffer  byte buffer storing the key.
+	 * @param keyOffset      offset of the key.
+	 * @param keyLen         length of the key.
+	 * @param returnOldState whether to return old state.
+	 * @return the old state. Null will be returned if key does not exist or returnOldState is false.
+	 */
+	private S removeNode(ByteBuffer keyByteBuffer, int keyOffset, int keyLen, boolean returnOldState) {
+		int deleteCount = 0;
+		long prevNode = findPredecessor(keyByteBuffer, keyOffset, 1);
+		long currentNode = helpGetNextNode(prevNode, 0);
+		long nextNode;
+
+		int c;
+		while (currentNode != NIL_NODE) {
```

Comment:
I have made a commit to extract and unify the logic, but have some concerns about the performance impact because the refactor creates additional `Tuple3` and `Tuple4` objects on the very core code path.

Reply:
And I didn't extract an `Iterator` since the logic here cannot adapt well to iterator semantic. Please check the new commit for details. Thanks.